kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/Checkpointer.java

```diff
@@ -76,6 +77,92 @@ public static Optional<CheckpointInstance> getLatestCompleteCheckpointFromList(
         }
     }
 
+    /**
+     * Utility method to find the last complete checkpoint before a given version.
+     */
+    public static Optional<CheckpointInstance> findLastCompleteCheckpointBefore(
+            TableClient tableClient,
+            Path tableLogPath,
+            long version) {
+        return findLastCompleteCheckpointBeforeHelper(tableClient, tableLogPath, version)._1;
+    }
+
+    protected static Tuple2<Optional<CheckpointInstance>, Long>
+            findLastCompleteCheckpointBeforeHelper(
+                    TableClient tableClient,
+                    Path tableLogPath,
+                    long version) {
+        CheckpointInstance upperBoundCheckpoint = new CheckpointInstance(version);
+        logger.info("Try to find the last complete checkpoint before version {}", version);
+
+        // This is a just a tracker for testing purposes
+        long numberOfFilesSearched = 0;
+        long currentVersion = version;
+
+        // Some cloud storage APIs make a calls to fetch 1000 at a time.
+        // To make use of that observation and to avoid making more listing calls than
+        // necessary, list 1000 at a time (backwards from the given version). Search
+        // within that list if a checkpoint is found. If found stop, otherwise list the previous
+        // 1000 entries. Repeat until a checkpoint is found or there are no more delta commits.
+        while (currentVersion >= 0) {
+            try {
+                long searchLowerBound = Math.max(0, currentVersion - 1000);
+                CloseableIterator<FileStatus> deltaLogFileIter = tableClient.getFileSystemClient()
+                        .listFrom(FileNames.listingPrefix(tableLogPath, searchLowerBound));
+
+                List<CheckpointInstance> checkpoints = new ArrayList<>();
+                while (deltaLogFileIter.hasNext()) {
+                    FileStatus fileStatus = deltaLogFileIter.next();
+                    String fileName = new Path(fileStatus.getPath()).getName();
+
+                    long currentFileVersion;
+                    if (FileNames.isCommitFile(fileName)) {
+                        currentFileVersion = FileNames.deltaVersion(fileName);
+                    } else if (FileNames.isCheckpointFile(fileName)) {
+                        currentFileVersion = FileNames.checkpointVersion(fileName);
+                    } else {
+                        // allow all other types of files.
+                        currentFileVersion = Math.min(currentVersion, version);
```

Comment:
We already assign `long currentVersion = version;`.

Does `currentVersion` ever increase? If not, could we just do `currentFileVersion = currentVersion`